Ab/working on references

abarciauskas-bgse opened the pull request:
Adds:
* support for multiscale and references
* basic tests

Needs:
* tests for different parameters 
~ideally, support for tiling without an internet connection in tests~
* ~exploration into decode_times - do we get valid responses without decoding times? Without passing `decode_times=False` we get~
```python
ValueError: unable to decode time units 'days since 1981-01-31' with "calendar 'proleptic_gregorian'". Try opening your dataset with decode_times=False or installing cftime if it is not installed.
```

vincentsarago commented on test_factory.py:
```diff
@@ -0,0 +1,105 @@
+        "variable": "sst",
+        "reference": True
+    })
+    print(response.json())
```
```suggestion

```

abarciauskas-bgse commented on test_factory.py:
```diff
@@ -0,0 +1,104 @@
+    assert response.headers['Content-Type'] == 'image/png'
+
+def test_get_tile():
+    power_zarr = "s3://power-analysis-ready-datastore/power_901_monthly_meteorology_utc.zarr"
```
@vincentsarago yes, will work on that today if I have time

abarciauskas-bgse commented:
@vincentsarago I reworked the tests to use netcdfs, reference files and zarr stores generated locally.

FWIW I wanted to understand a bit better what is the implication of allowing for `decode_times=False` which seems to be required to open the dataset if the time values cannot be decoded to numpy dtypes datetime64[ns] and timedelta64[ns] and it appears that, if times are not decoded, we will not be able to use a lot of the datetime utilities offered by numpy and pandas. For now, I don't think we need those utilities in our current routes and if we end up implementing those utilities then `decode_times=False` will be an incompatible parameter value.

abarciauskas-bgse commented:
Do you think we should add more tests or can I open mark this PR as ready for review?

vincentsarago commented:
> Do you think I should add more tests or can I open mark this PR as ready for review? 

I think it's fine for now 🙏

vincentsarago commented on titiler/xarray/factory.py:
```diff
@@ -1,6 +1,8 @@
 
 from dataclasses import dataclass
+import fsspec
+import s3fs
```
is `s3fs` used anywhere?

vincentsarago commented on titiler/xarray/factory.py:
```diff
@@ -18,6 +20,78 @@
 from titiler.core.resources.responses import JSONResponse
 
 
+def xarray_open_dataset(
```
move them outside the factory for now, but we can add them back later as dependencies

abarciauskas-bgse replied:
sorry - do you mean add these functions outside of the factory, to another module, and then import them here?

vincentsarago replied:
for now we can keep them in `factory.py` but there is no real need to have them in the `XarrayTilerFactory`

vincentsarago commented on requirements-dev.txt:
```diff
@@ -5,3 +5,17 @@ constructs>=10.0.0
+s3fs==2023.4.0
+titiler==0.11.6
+xarray==0.19.0
+zarr==2.14.2
```
requirements-devs aimed to be used only for deployment purposes. 

if we need more `dev` dependencies for the `titiler.xarray` package we should update the pyproject.toml

abarciauskas-bgse replied:
addressing this now!

vincentsarago commented on titiler/xarray/factory.py:
```diff
@@ -18,6 +20,78 @@
 
+def xarray_open_dataset(
+    src_path: str,
+    z: Optional[str] = None,
```
Z is an optional `String` but in the factory it's defined as an integer

abarciauskas-bgse replied:
fixed

vincentsarago commented on pyproject.toml:
```diff
@@ -25,16 +25,21 @@ classifiers = [
-    "fsspec",
-    "s3fs",
+    "fsspec==2023.4.0",
+    "s3fs==2023.4.0",
```
😬 do we really need all those dependencies? 

This adds a lot

vincentsarago replied:
e.g we don't need to add 
```
    "aiobotocore==2.5.0",
     "botocore==1.29.76",
     "mangum>=0.10.0",
``` 

here because those dependency are only useful in AWS Lambda (which we install in the docker image)

abarciauskas-bgse replied:
I'll see which ones we need, I want to make sure the dependencies installed locally are the same versions as get run in the docker image so I can hope that if my local version is running the docker version will also work. A few times now things are broken in the deployed instance when things are working locally, due to dependency differences.

vincentsarago replied:
I think it's important to make the difference between what we use locally and what we have in the AWS Lambda environment. E.g the version for `s3fs` will be entirely defined by the version of botocore found in AWS Lambda which IMO is not something that would have impact on running the code locally but might mess with your local environment (if you're not using virtual env)

abarciauskas-bgse replied:
I see what you are saying. I feel it is very useful to know how to run the application locally so that it simulates the deployed version and makes it easier to troubleshoot locally. As a middle ground, I have removed the extra dependencies from this file and write some about how to install AWS Lambda dependencies in the README. Let me know if this approach works for you.

vincentsarago replied:
we could totally adds a `requirement-lambda.txt` if you want! (my original comments was that we shouldn't add things to `-dev` because it was use for deployment purpose)

vincentsarago replied:
I'll refactor the whole `stack` directory to have something more like https://github.com/developmentseed/eoAPI/tree/main/infrastructure in another PR

vincentsarago commented on pyproject.toml:
```diff
@@ -25,16 +25,21 @@ classifiers = [
+    "mangum>=0.10.0",
+    "netCDF4==1.6.2",
+    "xarray==0.19.0",
+    "rioxarray==0.14.1",
```
what's is the motivation for pinning to exact version?

vincentsarago commented on package.json:
```diff
@@ -5,6 +5,7 @@
   "license": "MIT",
   "private": true,
   "dependencies": {
+    "aws-cdk": "^2.76.0",
```
?

abarciauskas-bgse replied:
I found this needed to be the same version as in requirements-dev.txt or I get an error about mismatching versions. But perhaps this is only if the virtualenv is not setup correctly, I will check.

vincentsarago replied:
interesting! 

the `cdk` depends on `aws-cdk` so my understanding is that when installing `cdk` it should install the right `aws-cdk` (https://www.npmjs.com/package/cdk?activeTab=code)

abarciauskas-bgse replied:
removed

abarciauskas-bgse commented:
Thanks @vincentsarago for all your help - I think I have addressed your comments and the CI tests are passing. I think we can postpone including a requirements-lambda.txt for now!

vincentsarago commented on README.md:
````diff
@@ -1,23 +1,48 @@
-# Install node dependencies
-npm install
+```bash
+pytest
````
```suggestion
python -m pytest --cov titiler.xarray --cov-report term-missing -s -vv
```

vincentsarago commented on README.md:
```diff
@@ -1,23 +1,48 @@
-# Create Virtual env and install python dependencies
-python -m pip install --upgrade virtualenv
-virtualenv .venv
+Tests use data generated locally by using `fixtures/generate_test_*.py` scripts.
```
```suggestion
Tests use data generated locally by using `tests/fixtures/generate_test_*.py` scripts.
```

vincentsarago commented on README.md:
````diff
@@ -1,23 +1,48 @@
+To run just one test:
 
+```bash
+pytest test_factory.py::test_get_info
````
```suggestion
python -m pytest tests/test_app.py::test_get_info --cov titiler.xarray --cov-report term-missing -s -vv
```

vincentsarago reviewed (approved):
🚢 🚀
